### PR TITLE
Adding EQDSK support 

### PR DESCRIPTION
**Pull request by fusionby2030**
A draft for EQDSK support by modifying the geometry, geometry_loader, and build_sim files. 

A corresponding geometry file is contributed to `data/third_party/geo` (credit to Antti Salmi for the help in geometry file generations). 

Additionally, an example configuration is made based off  (copied from)  the iter hybrid scenario and basic example already provided.  But the PR can likely do without the `eqdsk_example` config... the example takes a long time to run (at least it runs :D)...

The flux surface averaging is done using contourpy and RectBivariate spline interpolation of the PSI 2D grid. Note, there appear to be systematic differences in the shaping parameters (R inboard, R outboard, triangularity).

Somewhere the requirements have to be updated to include the [TSVV-15's eqdsk](https://github.com/Fusion-Power-Plant-Framework/eqdsk) and [contourpy](https://pypi.org/project/contourpy/)

See #373 for more details and discussion.

### CODE REVIEW

File: torax/geometry.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 import jax.numpy as jnp
 import numpy as np
 import scipy
+from contourpy import contour_generator
```
Comment (tamaranorman) [inline torax/geometry.py:31]: Can you add new dependencies to here

https://github.com/google-deepmind/torax/blob/main/pyproject.toml

Also can you make it such that the dependencies are ordered alphabetically - although keeping the groupings

File: torax/geometry.py [hunk 2]
```diff
@@ -879,6 +880,213 @@ def from_fbt(
...
+  def from_eqdsk(
+        cls,
+        geometry_dir: str | None = None,
+        geometry_file: str = 'eqdsk_cocos02.eqdsk',
+        hires_fac: int = 4,
+        Ip_from_parameters: bool = True,
+        n_rho: int = 25,
+        # NOTE: The other geometry factors are coming from eqdsk...
+    ):
+        # helper function to calculate area of a polygon given curve along x, z
```
Comment (tamaranorman) [inline torax/geometry.py:893]: Use """ method comments for method level comments

And also add types

File: torax/geometry.py [hunk 3]
```diff
@@ -879,6 +880,213 @@ def from_fbt(
...
+        def calculate_area(x, z):
+            # Gauss-shoelace formula (https://en.wikipedia.org/wiki/Shoelace_formula)
+            # could likely use the determinant approach
+            #  0.5 * np.abs(np.dot(x, np.roll(z, 1)) - np.dot(z, np.roll(x, 1)))
```
Comment (tamaranorman) [inline torax/geometry.py:897]: Is this an alternative method, if we aren't using this I am not sure we need to add this comment

File: torax/geometry_loader.py [hunk 4]
```diff
@@ -58,6 +59,11 @@ def _load_fbt_data(file_path: str) -> dict[str, np.ndarray]:
   """Loads the data from a FBT-LY file into a dictionary."""
   return scipy.io.loadmat(file_path, squeeze_me=True)
 
+from eqdsk import EQDSKInterface
```
Comment (tamaranorman) [inline torax/geometry_loader.py:62]: Move imports to the top and also add to dependency file

File: torax/geometry_loader.py [hunk 5]
```diff
@@ -58,6 +59,11 @@ def _load_fbt_data(file_path: str) -> dict[str, np.ndarray]:
...
+def _load_eqdsk_data(file_path: str) -> dict[str, np.ndarray]:
+    eq = EQDSKInterface.from_file(file_path, no_cocos=True) # should probably handle COCOS shenanigans...
```
Comment (tamaranorman) [inline torax/geometry_loader.py:64]: Is this something we need to handle here? Or maybe turn it into a formal todo

File: torax/geometry_loader.py [hunk 6]
```diff
@@ -58,6 +59,11 @@ def _load_fbt_data(file_path: str) -> dict[str, np.ndarray]:
...
+    return eq.__dict__
```
Comment (tamaranorman) [inline torax/geometry_loader.py:65]: Use dict(eq) but also maybe rename to something like loaded_data or similar as eq is generally used for equals

File: torax/geometry.py [hunk 7]
```diff
@@ -879,6 +880,213 @@ def from_fbt(
...
+        # --- Area, Volume, R_inboard, R_outboard
+        # --- FSA: <1/R^2>, <Bp^2>, <|grad(psi)|>, <|grad(psi)|^2>
+        # --- Toroidal plasma current
+        # --- Integral dl/Bp
+        # -----------------------------------------------------------
+
+        # Gathering area for profiles
+        areas, volumes                = [], []
+        R_inboard, R_outboard         = [], []
+        flux_surf_avg_1_over_R2_eqdsk = [] # <1/R**2>
```
Comment (tamaranorman) [inline torax/geometry.py:966]: nit: the underscore eqdsk is probably not necessary here?

To save vertical space etc is it worth preallocating the numpy arrays and then assigning in rather than needing to create the arrays and then turn them into arrays?

**Review by tamaranorman — CHANGES_REQUESTED** [review]
Thanks for the PR it looks really good and adds some nice functionality just a few small mainly formatting etc comments

File: torax/geometry_loader.py (hunk 5 above)
Comment (fusionby2030) [reply torax/geometry_loader.py]: This needs a formal TODO, not sure how the torax team wants to handle COCOS in the future.

File: torax/geometry_loader.py (hunk 6 above)
Comment (fusionby2030) [reply torax/geometry_loader.py]: dict(eq) did not work, therefore I have kept it as eq.__dict, except renaming as suggested.

File: torax/geometry.py (hunk 7 above)
Comment (fusionby2030) [reply torax/geometry.py]: I think the underscore is useful as the `from_chease` method has similar syntax for the FSA quantities.

**Comment by tamaranorman** [comment]
Looks good, are you able to resolve the conflicts so that we can try it out

**Review by tamaranorman — APPROVED** [review]

File: torax/geometry.py [hunk 8]
```diff
@@ -879,6 +880,197 @@ def from_fbt(
...
+        offset = 0.01
+        mask = (X > Xlcfs.min() - offset) & (X < Xlcfs.max() + offset) & (Z > Zlcfs.min() - offset) & (Z < Zlcfs.max() + offset)
+        masked_psi_eqdsk_2dgrid = np.ma.masked_where(~mask, psi_eqdsk_2dgrid)
+
+        # q on uniform grid (pressure, etc., also defined here)
+        q_eqdsk_1dgrid = eqfile['qpsi']
+
+        # ---- Interpolations
+        q_interp = scipy.interpolate.interp1d(psi_eqdsk_1dgrid, eqfile['qpsi'], kind='cubic')
+        psi_spline_fit   = scipy.interpolate.RectBivariateSpline(X_1D, Z_1D, psi_eqdsk_2dgrid, kx=3, ky=3, s=0)
```
Comment (jcitrin) [inline torax/geometry.py:932]: out of curiosity, did you check robustness of the results to the hyperparameters of this method?

File: torax/geometry.py [hunk 9]
```diff
@@ -879,6 +880,197 @@ def from_fbt(
...
+        F_interp = scipy.interpolate.interp1d(psi_eqdsk_1dgrid, eqfile['fpol'], kind='cubic') # toroidal field flux function
+
+        # -----------------------------------------------------------
+        # --------- Make flux surface contours ---------
+        # -----------------------------------------------------------
+
+        # set epsilon based on floating point precision of psi grid
+        # NOTE: this is hacky and should be replaced with a more robust method
+        epsilon = 1e-7
```
Comment (jcitrin) [inline torax/geometry.py:941]: nit: can use eps from the `constants.Constants` dataclass

File: torax/data/third_party/geo/eqdsk_cocos02.eqdsk
Comment (jcitrin) [inline torax/data/third_party/geo/eqdsk_cocos02.eqdsk:1]: Was this generated directly from the existing ITER_hybrid CHEASE file?

If so, and if it doesn't break the EQDSK file, would be good to mention this in a header

File: torax/examples/eqdsk_example.py
Comment (jcitrin) [inline torax/examples/eqdsk_example.py:1]: if all else apart from geometry is identical to test_iterhybrid_predictorcorrector, then can shorten this file as well as making the provenance clearer, by copying the CONFIG from iterhybrid and then modifying it. See https://github.com/google-deepmind/torax/blob/main/torax/tests/test_data/test_iterhybrid_rampup_restart.py for an example

**Comment by jcitrin** [comment]
Thanks, looks really good. Made some minor comments after another run through.

After these, and merging with main, we'll bring it in internally and push it from there, next week. Thanks again for this!

File: torax/geometry.py (hunk 8 above)
Comment (fusionby2030) [reply torax/geometry.py:932]: IIRC, I did do some testing with this, but determined the resolution of the underlying eqdsk file is more important than the hyperparameters of the spline fitting. Perhaps a warning is of use for this.

File: torax/examples/eqdsk_example.py
Comment (fusionby2030) [reply torax/examples/eqdsk_example.py:1]: Replaced with the example. That is much cleaner.

**Comment by theo-brown** [comment]
Is this PR going to be merged in any time soon? Keen to use it to smooth out the JETTO->TORAX workflow.

**Comment by jcitrin** [comment]
@fusionby2030 , @theo-brown , this is now merged with [f7082a1](https://github.com/google-deepmind/torax/commit/f7082a1c503cd2fea8449606719632b9490d9712)

Some changes compared to the last PR commit:

1. Modification of minor radius definition from EQDSK, to the midplane average instead of distance from axis to LFS

2. A sim integration test was set up, with appropriate parameters for the EQDSK small-tokamak-equilibrium that was generated for the demo. @fusionby2030 , do you have further information on what this equilibrium was based on? We can add this info in a small PR. In any case it runs quite fast. The issue with the slow running sim beforehand was that the iterhybrid config was not appropriate for this EQDSK file. e.g. very high Ip compared to the EQDSK B0, leading to non-physical q<<1 , as well as extremely high power densities. This led to QLKNN to go haywire and produced chis at the max towards 100, which then led to small timesteps due to the chi time_step_calculator, which is adaptive and sets t \propto 1/chi.
  
3. Fixing minor linter issues. Recommend using pylint. 

Apologies for not being more transparent with pushing to this branch and discussing these changes. Next time when collaborating on the github version, will concentrate on carrying out the review and pushing iterative changes on github, as opposed to our internal version.

Thanks again @fusionby2030 for this very valuable feature! @theo-brown , looking forward to benchmarks against JETTO-EQDSK. Would be happy to discuss going forward.

**Comment by theo-brown** [comment]
Agreed, many thanks @fusionby2030 - a major contribution to the tool!

**Comment by jcitrin** [comment]
@theo-brown The CHEASE iterhybrid equivalent EQDSK landed in pint. I'm bringing it now into TORAX and will take a first look at the EQDSK-CHEASE comparison.
